crates/binjs_io/src/entropy/read/prelude_decoders.rs

```diff
@@ -162,26 +162,14 @@ where
         debug!(target: "read", "SectionDecoder::read_stream preparing to decompress {} bytes", byte_len);
 
         // Extract slice.
-        let mut buf = Vec::with_capacity(byte_len); // FIXME: We may want some maximal size here.
-        unsafe {
-            buf.set_len(byte_len);
-        }
+        let mut buf = vec![0; byte_len]; // Note: We may wish to impose a maximal size here.
         self.input
             .read_exact(buf.as_mut_slice())
             .map_err(TokenReaderError::ReadError)?;
 
-        // Note: This is the only way I have found to get Brotli to decompress all the
-        // bytes we feed it.
-        use std::io::Write;
-
         let mut result = Vec::new();
-        {
-            let mut brotli = brotli::DecompressorWriter::new(&mut result, 32768); // FIXME: WTF? This doesnt seem to work with small buffer sizes. How am I supposed to guess the buffer size?
-            brotli
-                .write_all(&buf)
-                .map_err(TokenReaderError::ReadError)?;
-            brotli.flush().map_err(TokenReaderError::ReadError)?;
-        }
+        brotli::BrotliDecompress(&mut Cursor::new(buf), &mut result)
```

Comment:
On the second thought... why do we need temporary buffer here? Can we pass slice of input directly to Brotli?

I imagine passing something like `self.input.take(byte_len)` should work (maybe with some casting to avoid consuming `self.input` but rather `&mut self.input` though, or maybe Rust will be smart enough to figure this out itself).

Reply:
Yes, we could use `self.input.take`, which would save some copies and allocations, but in case of encoding bug, it's going to make it harder for use to find the issue. So, if you don't mind, I'd rather keep the current implementation.

Reply:
would be nice to add comment about that optimization opportunity.

Reply:
Ok, I have added a workaround that should bring us the best of both worlds.

Reply:
Hmm, I'm not sure how is `BoundedRead` different from what `take` does. Could you add some comments please?

Reply:
> Ah are you saying that `BrotliDecompress` silently ignores extra data in the end of input? If so, that's unfortunate...

I don't want to rely upon this (unspecified) behavior. I want to be able to check that the length is exactly the one that was indicated in the header.

> But maybe a simpler solution would be to provide util that only tracks amount of data read, and then compose that with `.take()` instead of reimplementing `.take()` logic too? I believe it should be a bit simpler.

I already had `Pos` at hand, it seemed simpler to reuse it.

Reply:
> I already had `Pos` at hand, it seemed simpler to reuse it.

What is `Pos`? My suggestion is to change the helper to only track position after each `read`, and not check for EOF etc. (leave that to built-in `take`). That should significantly simplify the logic and allow to reuse it in more contexts.

Reply:
Done.